Source/WebCore/platform/spiel/PlatformSpeechSynthesizerSpiel.cpp

```diff
@@ -0,0 +1,258 @@
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "PlatformSpeechSynthesizer.h"
+
+#if ENABLE(SPEECH_SYNTHESIS) && USE(SPIEL)
+
+#include "GStreamerCommon.h"
+#include "PlatformSpeechSynthesisUtterance.h"
+#include "PlatformSpeechSynthesisVoice.h"
+#include "WebKitAudioSinkGStreamer.h"
+#include <spiel/spiel.h>
+#include <wtf/TZoneMallocInlines.h>
+#include <wtf/text/MakeString.h>
+
+namespace WebCore {
+
+class SpielSpeechWrapper {
+    WTF_MAKE_TZONE_ALLOCATED_INLINE(SpielSpeechWrapper);
+    WTF_MAKE_NONCOPYABLE(SpielSpeechWrapper);
+public:
+    explicit SpielSpeechWrapper(const PlatformSpeechSynthesizer&);
+    ~SpielSpeechWrapper();
+
+    Vector<RefPtr<PlatformSpeechSynthesisVoice>> initializeVoiceList();
+    void pause();
+    void resume();
+    void speakUtterance(RefPtr<PlatformSpeechSynthesisUtterance>&&);
+    void cancel();
+    void resetState();
+
+private:
+    RefPtr<PlatformSpeechSynthesisUtterance> m_utterance;
+    const PlatformSpeechSynthesizer& m_platformSynthesizer;
+    GRefPtr<SpielSpeaker> m_speaker;
+    Vector<GRefPtr<SpielVoice>> m_voices;
+};
+
+SpielSpeechWrapper::SpielSpeechWrapper(const PlatformSpeechSynthesizer& synthesizer)
+    : m_platformSynthesizer(synthesizer)
+{
+    ensureGStreamerInitialized();
+    registerWebKitGStreamerElements();
+
+    GRefPtr<GstElement> audioSink = createPlatformAudioSink("speech"_s);
+    if (!audioSink) {
+        GST_ERROR("Failed to create GStreamer audio sink element");
+        return;
+    }
+
+    // Async API can't really be used here because the voices initialization
+    // (PlatformSpeechSynthesizer::initializeVoiceList()) is synchronous.
+    GUniqueOutPtr<GError> error;
+    m_speaker = adoptGRef(spiel_speaker_new_sync(nullptr, &error.outPtr()));
```

Comment:
A common pattern is to initialize speech in an async fashion and return an empty list for voices. Once it is initialized you call `voicesDidChange` to notify content that voices are available.

Reply:
The demo I tried this with is https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/

Reply:
That is strange since the [script.js](https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/script.js) does listen voice "voiceschanged" to re-populate the dropdown. I'm curious if this indeed does get emitted for you. Maybe getVoices is retuning null instead of an empty list so the script fails?

Reply:
That is strange since the [script.js](https://mdn.github.io/dom-examples/web-speech-api/speak-easy-synthesis/script.js) does listen voice "voiceschanged" to re-populate the dropdown. I'm curious if this indeed does get emitted for you. Maybe getVoices is retuning null instead of an empty list so the script fails?

Reply:
This is what I see in a first time run of Chrome:
![image](https://github.com/user-attachments/assets/039f314c-924f-44df-8780-613c76cfea57)

This is the URL:
`data:text/html,<script>function populateVoices() {let voices = speechSynthesis.getVoices();console.log('Voice count', voices.length);}populateVoices();speechSynthesis.onvoiceschanged = populateVoices;</script>`

Reply:
Got it working with async :)